### ISSUE TITLE
Refactor SMTP Services to be more object based

### ISSUE THREAD

**Issue by rjhowell44** [prominenceai/deepstream-services-library#471]
The SMTP Services need to support the creation of a Mailer Object that can then be added to one or more Recording Taps, Recording Sinks, and Image Capture Actions -- the same as how Image and Video Players are added to these components directly.

On Recording and Capture complete, the components will use the Mailer Object to send out an email message with the details and location of the new file. Image Capture actions can (optionally) include the image in the email. 

If the components own an Image or Video RTSP Player (once implemented) - to stream the image or video to the network - the email will include the URI for the outgoing stream as well

**Comment by rjhowell44** [comment]
New proposed API

New `MAILER` object return types
```C
#define DSL_RESULT_MAILER_RESULT                                    0x00500000
#define DSL_RESULT_MAILER_NAME_NOT_UNIQUE                           0x00500001
#define DSL_RESULT_MAILER_NAME_NOT_FOUND                            0x00500002
#define DSL_RESULT_MAILER_THREW_EXCEPTION                           0x00500003
#define DSL_RESULT_MAILER_IN_USE                                    0x00500004
#define DSL_RESULT_MAILER_SET_FAILED                                0x00500005
#define DSL_RESULT_MAILER_PARAMETER_INVALID                         0x00500006
```

New SMTP Mailer Constructor
```C
/**
 * @brief Creates a uniquely named SMTP Mailer 
 * @param[in] name unique name for the new Mailer
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_new(const wchar_t* name);
```
All existing `dsl_smtp_<*>` methods changed to `dsl_mailer_<*>` with the inclusion of a `name` parameter to name the object to use.
 
```C
/**
 * @brief Gets the current Enabled state for the named SMTP Mailer
 * @param[in] name unique name of the Mailer to query
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_enabled_get(const wchar_t* name, boolean* enabled);

/**
 * @brief Sets the Enabled state of the named SMTP Mailer
 * Disabling the Mailer will block all subsequent emails from being queued for sending.
 * @param[in] name unique name of the Mailer to update
 * @param[in] enabled set to true to enable, false to disabled
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_enabled_set(const wchar_t* name, boolean enabled);

/**
 * @brief Sets the user credentials for the Mailer's SMTP host for all subsequent emails
 * @param[in] name unique name of the Mailer to update
 * @param[in] username username to use
 * @param[in] password password to use
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_credentials_set(const wchar_t* name, 
    const wchar_t* username, const wchar_t* password);

/**
 * @brief Gets the current SMTP server URL setting for the named Mailer
 * @param[in] name unique name of the Mailer to query
 * @param[out] server_url current server URL in use
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_server_url_get(const wchar_t* name,
    const wchar_t** server_url);

/**
 * @brief Sets the SMTP server URL for the Mailer to use for all 
 * subsequence email sent out
 * @param[in] name unique name of the Mailer to update
 * @param[in] server_url to use 
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_server_url_set(const wchar_t* name,
    const wchar_t* server_url);

/**
 * @brief Gets the current From address for the named Mailer
 * @param[in] name unique name of the Mailer to query
 * @param[out] display_name current From address display name
 * @param[out] address current From address
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_address_from_get(const wchar_t* name,
    const wchar_t** display_name, const wchar_t** address);

/**
 * @brief Sets the current From address for the Mailer to use 
 * for all subsequent emails
 * @param[in] name unique name of the Mailer to update
 * @param[in] display_name new From address display name to use
 * @param[in] address new From address
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_address_from_set(const wchar_t* name,
    const wchar_t* display_name, const wchar_t* address);

/**
 * @brief Returns the current SMTP SSL enabled setting in use by the named Mailer
 * The setting is enabled by default
 * @param[in] name unique name of the Mailer to query
 * @param[out] enabled true if SSL is enabled, false otherwise 
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
 */
DslReturnType dsl_mailer_ssl_enabled_get(const wchar_t* name, boolean* enabled);

/**
 * @brief Sets the SMTP SSL enabled setting for the named Mailer
 * @param[in] name unique name of the Mailer to update
 * @param[in] enabled set to true to enable, false otherwise
 * @return DSL_RESULT_SUCCESS on success, one of DSL_MAILER_RESULT otherwise
*/
DslReturnType dsl_mailer_ssl_enabled_set(const wchar_t* name, boolean enabled);

/**
 * @brief Adds a new email address to the To list of the named Mailer
 * @param[in] name unique name of the Mailer to update
 * @param[in] display_name display name for the To address
 * @param[in] address qualifed email To address
 * @return DSL_RESULT_SUCCESS on success, one of DSL_RESULT otherwise.
 */
DslReturnType dsl_mailer_address_to_add(const wchar_t* name,
    const wchar_t* display_name, const wchar_t* address);

/**
 * @brief Removes all current To addresses from the named Mailer
 * @param[in] name unique name of the Mailer to update
 * @return DSL_RESULT_SUCCESS on success, one of DSL_RESULT otherwise.
 */
DslReturnType dsl_mailer_address_to_remove_all(const wchar_t* name);

/**
 * @brief Adds a new email address to the Cc list of the name Mailer
 * @param[in] name unique name of the Mailer to update
 * @param[in] display_name display name for the Cc address
 * @param[in] address qualifed email Cc address
 * @return DSL_RESULT_SUCCESS on success, one of DSL_RESULT otherwise.
 */
DslReturnType dsl_mailer_address_cc_add(const wchar_t* name,
    const wchar_t* display_name, const wchar_t* address);

/**
 * @brief Removes all current CC addresses from the named Mailer
 * @param[in] name unique name of the Mailer to update
 * @return DSL_RESULT_SUCCESS on success, one of DSL_RESULT otherwise.
 */
DslReturnType dsl_mailer_address_cc_remove_all(const wchar_t* name);

/**
 * @brief Sends a test message using the current SMTP
 * settings and email addresses (From, To, Cc)
 * @param[in] name unique name of the Mailer to test
 * @return DSL_RESULT_SUCCESS on success, one of DSL_RESULT otherwise.
 */
DslReturnType dsl_mailer_test_message_send(const wchar_t* name);
```

**Comment by rjhowell44** [comment]
Done. Merged into v0.14.alpha
